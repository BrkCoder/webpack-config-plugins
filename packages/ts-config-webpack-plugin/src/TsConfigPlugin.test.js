const fs = require('fs');
const path = require('path');
const rimraf = require('rimraf');
const webpack = require('webpack');
const TsConfigPlugin = require('./TsConfigPlugin');

// declare global test timeout to 10s for async tests
jest.setTimeout(10 * 1000);

const runWithinWebpackContext = (callback, settings = {}) => {
	webpack({
		mode: settings.mode || 'development',
		entry: path.resolve(__dirname, 'fixtures/entry.ts'),
		target: 'node',
		node: {
			global: false,
		},
		output: {
			path: path.resolve(__dirname, 'fixtures/dist'),
			filename: 'bundle.js',
			libraryTarget: 'umd',
		},
		plugins: [new TsConfigPlugin(settings.pluginOptions)],
	}).run((_, stats) => {
		callback({
			_,
			stats,
			options: stats.compilation.options,
			errors: stats.compilation.errors,
		});
	});
};

beforeEach(() => {
	jest.resetModules();
});

afterEach(resolve => {
	rimraf(path.join(__dirname, './fixtures/dist'), resolve);
});

describe('TsConfigPlugin', () => {
	it('should be creatable without options', () => {
		new TsConfigPlugin();
	});

	describe('runnable webpack context', () => {
		it('should finish without errors', () => {
			runWithinWebpackContext(({ errors }) => {
				expect(errors).toEqual([]);
			});
		});

		it('options should match the last options snapshot', resolve => {
			runWithinWebpackContext(({ options }) => {
				expect(options).toMatchSnapshot();
				resolve();
			});
		});

		it('should set plugins correctly', resolve => {
			runWithinWebpackContext(({ options }) => {
				expect(options.plugins.length).toBe(2);
				expect(options.plugins[0] instanceof TsConfigPlugin).toBe(true);
				resolve();
			});
		});

		it('should set rules correctly', resolve => {
			runWithinWebpackContext(({ options }) => {
				expect(options.module.rules.length).toBe(1);
				expect(options.module.rules[0].test.test('entry.ts')).toBe(true);
				resolve();
			});
		});

		it(
			'should generate correct output',
			() => {
				runWithinWebpackContext(() => {
					const bundlePath = path.resolve(__dirname, './fixtures/dist/bundle.js');
					const contents = fs.readFileSync(bundlePath).toString();
					expect(contents).toBeDefined();
				});
			},
			{ mode: 'production' }
		);
	});
});
