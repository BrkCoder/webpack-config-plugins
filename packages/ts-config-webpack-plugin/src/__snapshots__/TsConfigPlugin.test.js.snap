// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TsConfigPlugin runnable webpack context options should match the last options snapshot 1`] = `
Object {
  "cache": true,
  "context": "/Users/jbiasi/Projects/Namics OpenSource/webpack-config-plugins/packages/ts-config-webpack-plugin",
  "devtool": "cheap-eval-source-map",
  "entry": "/Users/jbiasi/Projects/Namics OpenSource/webpack-config-plugins/packages/ts-config-webpack-plugin/src/fixtures/entry.ts",
  "mode": "development",
  "module": Object {
    "defaultRules": Array [
      Object {
        "resolve": Object {},
        "type": "javascript/auto",
      },
      Object {
        "resolve": Object {
          "mainFields": Array [
            "main",
          ],
        },
        "test": /\\\\\\.mjs\\$/i,
        "type": "javascript/esm",
      },
      Object {
        "test": /\\\\\\.json\\$/i,
        "type": "json",
      },
      Object {
        "test": /\\\\\\.wasm\\$/i,
        "type": "webassembly/experimental",
      },
    ],
    "exprContextCritical": true,
    "exprContextRecursive": true,
    "exprContextRegExp": false,
    "exprContextRequest": ".",
    "rules": Array [
      Object {
        "test": /\\.\\(tsx\\?\\|d\\.ts\\)\\$/,
        "use": Array [
          Object {
            "loader": "/Users/jbiasi/Projects/Namics OpenSource/webpack-config-plugins/packages/ts-config-webpack-plugin/node_modules/cache-loader/dist/cjs.js",
            "options": Object {
              "cacheDirectory": "/Users/jbiasi/Projects/Namics OpenSource/webpack-config-plugins/packages/ts-config-webpack-plugin/node_modules/.cache-loader",
            },
          },
          Object {
            "loader": "/Users/jbiasi/Projects/Namics OpenSource/webpack-config-plugins/packages/ts-config-webpack-plugin/node_modules/thread-loader/dist/cjs.js",
            "options": Object {
              "workers": 7,
            },
          },
          Object {
            "loader": "/Users/jbiasi/Projects/Namics OpenSource/webpack-config-plugins/packages/ts-config-webpack-plugin/node_modules/ts-loader/index.js",
            "options": Object {
              "experimentalWatchApi": true,
              "happyPackMode": true,
              "transpileOnly": true,
            },
          },
        ],
      },
    ],
    "strictExportPresence": false,
    "strictThisContextOnImports": false,
    "unknownContextCritical": true,
    "unknownContextRecursive": true,
    "unknownContextRegExp": false,
    "unknownContextRequest": ".",
    "unsafeCache": true,
    "wrappedContextCritical": false,
    "wrappedContextRecursive": true,
    "wrappedContextRegExp": /\\.\\*/,
  },
  "node": Object {
    "Buffer": true,
    "__dirname": "mock",
    "__filename": "mock",
    "console": false,
    "global": false,
    "process": true,
    "setImmediate": true,
  },
  "optimization": Object {
    "checkWasmTypes": false,
    "concatenateModules": false,
    "flagIncludedChunks": false,
    "mangleWasmImports": false,
    "mergeDuplicateChunks": true,
    "minimize": false,
    "minimizer": Array [
      Object {
        "apply": [Function],
      },
    ],
    "namedChunks": true,
    "namedModules": true,
    "noEmitOnErrors": false,
    "nodeEnv": "development",
    "occurrenceOrder": false,
    "portableRecords": false,
    "providedExports": true,
    "removeAvailableModules": true,
    "removeEmptyChunks": true,
    "runtimeChunk": undefined,
    "sideEffects": false,
    "splitChunks": Object {
      "automaticNameDelimiter": "~",
      "cacheGroups": Object {
        "default": Object {
          "minChunks": 2,
          "priority": -20,
          "reuseExistingChunk": true,
        },
        "vendors": Object {
          "priority": -10,
          "test": /\\[\\\\\\\\\\\\/\\]node_modules\\[\\\\\\\\\\\\/\\]/,
        },
      },
      "chunks": "async",
      "maxAsyncRequests": Infinity,
      "maxInitialRequests": Infinity,
      "minChunks": 1,
      "minSize": 10000,
      "name": true,
    },
    "usedExports": false,
  },
  "output": Object {
    "chunkCallbackName": "webpackChunk",
    "chunkFilename": "[id].bundle.js",
    "chunkLoadTimeout": 120000,
    "crossOriginLoading": false,
    "devtoolLineToLine": false,
    "devtoolNamespace": "",
    "filename": "bundle.js",
    "globalObject": "global",
    "hashDigest": "hex",
    "hashDigestLength": 20,
    "hashFunction": "md4",
    "hotUpdateChunkFilename": "[id].[hash].hot-update.js",
    "hotUpdateFunction": "webpackHotUpdate",
    "hotUpdateMainFilename": "[hash].hot-update.json",
    "jsonpFunction": "webpackJsonp",
    "jsonpScriptType": false,
    "library": "",
    "libraryTarget": "umd",
    "path": "/Users/jbiasi/Projects/Namics OpenSource/webpack-config-plugins/packages/ts-config-webpack-plugin/src/fixtures/dist",
    "pathinfo": true,
    "sourceMapFilename": "[file].map[query]",
    "strictModuleExceptionHandling": false,
    "webassemblyModuleFilename": "[modulehash].module.wasm",
  },
  "performance": false,
  "plugins": Array [
    TsConfigPlugin {},
    ForkTsCheckerWebpackPlugin {
      "async": false,
      "cancellationToken": undefined,
      "checkDone": false,
      "checkSyntacticErrors": true,
      "colors": Chalk {
        "enabled": true,
      },
      "compilationDone": false,
      "compiler": undefined,
      "diagnostics": Array [],
      "doneCallback": [Function],
      "elapsed": undefined,
      "emitCallback": [Function],
      "formatter": [Function],
      "ignoreDiagnostics": Array [],
      "ignoreLints": Array [],
      "isWatching": false,
      "lints": Array [],
      "logger": CustomConsole {
        "_counters": Object {},
        "_formatBuffer": [Function],
        "_groupDepth": 0,
        "_timers": Object {},
        "assert": [Function],
        "clear": [Function],
        "count": [Function],
        "countReset": [Function],
        "debug": [Function],
        "dir": [Function],
        "error": [Function],
        "group": [Function],
        "groupCollapsed": [Function],
        "groupEnd": [Function],
        "info": [Function],
        "log": [Function],
        "time": [Function],
        "timeEnd": [Function],
        "trace": [Function],
        "warn": [Function],
        Symbol(counts): Map {},
        Symbol(groupIndent): "",
      },
      "memoryLimit": 2048,
      "options": Object {
        "async": false,
        "checkSyntacticErrors": true,
      },
      "silent": false,
      "started": undefined,
      "tsconfig": "./tsconfig.json",
      "tsconfigPath": undefined,
      "tslint": undefined,
      "tslintPath": undefined,
      "tslintVersion": undefined,
      "typescriptVersion": "2.9.2",
      "useColors": true,
      "vue": false,
      "watch": Array [],
      "watchPaths": Array [],
      "workersNumber": 1,
    },
  ],
  "resolve": Object {
    "aliasFields": Array [],
    "cacheWithContext": false,
    "extensions": Array [
      ".wasm",
      ".mjs",
      ".js",
      ".json",
    ],
    "mainFields": Array [
      "module",
      "main",
    ],
    "mainFiles": Array [
      "index",
    ],
    "modules": Array [
      "node_modules",
    ],
    "unsafeCache": true,
  },
  "resolveLoader": Object {
    "cacheWithContext": false,
    "extensions": Array [
      ".js",
      ".json",
    ],
    "mainFields": Array [
      "loader",
      "main",
    ],
    "mainFiles": Array [
      "index",
    ],
    "unsafeCache": true,
  },
  "target": "node",
}
`;
